training/src/main/scala/com/airbnb/aerosolve/training/KernelTrainer.scala

```diff
@@ -0,0 +1,126 @@
+    "rbf" -> FunctionForm.RADIAL_BASIS_FUNCTION,
+    "acos" -> FunctionForm.ARC_COSINE
+  )
+  
+  def train(sc : SparkContext,
+            input : RDD[Example],
+            config : Config,
+            key : String) : KernelModel = {
+    val modelConfig = config.getConfig(key)
+    val candidateSize : Int = modelConfig.getInt("num_candidates")
+
+    val learningRate : Float = Try(modelConfig.getDouble("learning_rate").toFloat).getOrElse(0.1f)
+    val rankKey : String = modelConfig.getString("rank_key")
+    val rankThreshold : Double = Try(modelConfig.getDouble("rank_threshold")).getOrElse(0.0f)
+
+    val examples = LinearRankerUtils
+        .makePointwiseFloat(input, config, key)
+    val model = initModel(modelConfig, examples)
+        
+    val loss : String = modelConfig.getString("loss")
+        
+    val candidates = examples.take(candidateSize)
+
+    // Super simple SGD trainer. Mostly to get the unit test to pass
+    for (candidate <- candidates) {
+      val gradient = computeGradient(model, candidate.example(0), loss, rankKey, rankThreshold)
+      if (gradient != 0.0) {
+        val flatFeatures = Util.flattenFeature(candidate.example(0));
+        model.onlineUpdate(gradient, learningRate, flatFeatures)
+      }
+    }
+   
+    model
+  }
+
+  def computeGradient(model : KernelModel,
+                      fv : FeatureVector,
+                      loss : String, rankKey : String, rankThreshold : Double) : Float = {
+    val prediction = model.scoreItem(fv)
+    val label = if (loss == "hinge") TrainingUtils.getLabel(fv, rankKey, rankThreshold) else TrainingUtils.getLabel(fv, rankKey)
+    loss match {
+      case "hinge" => {
+        val lossVal = scala.math.max(0.0, 1.0 - label * prediction)
+        if (lossVal > 0.0) {
+           return -label.toFloat
+        }
+      }
+      case "regression" => {
+        val diff = prediction - label
+        if (diff > 1.0) {
+          return 1.0f
+        }
+        if (diff < -1.0) {
+          return -1.0f
+        }
+      }
+    }
+    return 0.0f;
+  }
+
+  def initModel(modelConfig : Config, examples : RDD[Example]) : KernelModel = {
+    val kernel : String = modelConfig.getString("kernel")
+    val scale : Float = modelConfig.getDouble("scale").toFloat
+    val minCount : Int = modelConfig.getInt("min_count")
+    val rankKey : String = modelConfig.getString("rank_key")
+    val maxSV : Int = modelConfig.getInt("max_vectors")
+    
+    log.info("Building dictionary")
+    val dictionary = new StringDictionary();
+    val stats = TrainingUtils.getFeatureStatistics(minCount, examples)
+    log.info(s"Dictionary size is ${stats.size}")
+    
+    for (stat <- stats) {
+      val (family, feature) = stat._1
+      if (family != rankKey) {
+        val mean = stat._2.mean
+        val variance = Math.max(1e-6, stat._2.variance)
+        val scale = Math.sqrt(1.0 / variance)
+        dictionary.possiblyAdd(family, feature, mean, scale)
```

Comment:
Ok done also remember missing string features are 0 in the dense case so the variance is actually p * (1-p) where p is the probability of getting the string feature on. The variance is only 0 for the bias term as it is always on.